packages/migrations/src/actions/2023.01.17T10.46.28.import-legacy-s3-keys-to-database.mts

```diff
@@ -0,0 +1,613 @@
+  }
+
+  let lastCursor: null | Cursor = null;
+
+  do {
+    const items: zod.TypeOf<typeof TargetsModel> = await getPaginationTargets(lastCursor);
+    for (const item of items) {
+      const s3Key = `s3-legacy-keys/${item.id}`;
+      const privateAccessKey = generateLegacyCDNAccessToken(item.id);
+      const firstCharacters = privateAccessKey.substring(0, 3);
+      const lastCharacters = privateAccessKey.substring(privateAccessKey.length - 3);
+      const accessKeyHash = await bcryptjs.hash(privateAccessKey, await bcryptjs.genSalt());
+
+      // After creation on database
+      const response = await s3Client.fetch(
+        [env.S3_ENDPOINT, env.S3_BUCKET_NAME, s3Key].join('/'),
+        {
+          method: 'PUT',
+          body: accessKeyHash,
+        },
+      );
+
+      // Should probably do a retry here
+      if (response.status !== 200) {
+        throw new Error(`Unexpected Status for storing key. (status=${response.status})`);
+      }
+
+      const query = sql`
+      INSERT INTO
+        "cdn_access_tokens"
+        (
+          "target_id"
+          , "s3_key"
+          , "first_characters"
+          , "last_characters"
+          , "is_revoked"
+        )
+      VALUES
+        (
+          ${item.id}
+          , ${s3Key}
+          , ${firstCharacters}
+          , ${lastCharacters}
+          , false
+        )
+      ON CONFLICT DO NOTHING
+    `;
+
+      await connection.query(query);
+    }
+
+    lastCursor = null;
+    if (items.length > 0) {
+      lastCursor = {
+        lastId: items[items.length - 1].id,
+        lastCreatedAt: items[items.length - 1].created_at_cursor,
+      };
+    }
+  } while (lastCursor !== null);
+};
+
+export const down: slonik.Migration = async ({ context: { connection, sql } }) => {
+  await connection.query(sql`raise 'down migration not implemented'`);
+};
+
+const encoder = new TextEncoder();
+
+const HOST_SERVICES: Record<string, string | void> = {
+  appstream2: 'appstream',
+  cloudhsmv2: 'cloudhsm',
+  email: 'ses',
+  marketplace: 'aws-marketplace',
+  mobile: 'AWSMobileHubService',
+  pinpoint: 'mobiletargeting',
+  queue: 'sqs',
+  'git-codecommit': 'codecommit',
+  'mturk-requester-sandbox': 'mturk-requester',
+  'personalize-runtime': 'personalize',
+};
+
+// https://github.com/aws/aws-sdk-js/blob/cc29728c1c4178969ebabe3bbe6b6f3159436394/lib/signers/v4.js#L190-L198
+const UNSIGNABLE_HEADERS = new Set([
+  'authorization',
+  'content-type',
+  'content-length',
+  'user-agent',
+  'presigned-expires',
+  'expect',
+  'x-amzn-trace-id',
+  'range',
+  'connection',
+]);
+
+type AwsRequestInit = RequestInit & {
+  aws?: {
+    accessKeyId?: string;
+    secretAccessKey?: string;
+    sessionToken?: string;
+    service?: string;
+    region?: string;
+    cache?: Map<string, ArrayBuffer>;
+    datetime?: string;
+    signQuery?: boolean;
+    appendSessionToken?: boolean;
+    allHeaders?: boolean;
+    singleEncode?: boolean;
+  };
+};
+
+class AwsClient {
```

Comment:
any reason to copy the code here? we should be able to use deps, no?

Reply:
Not sure how it would work, to be honest. 😓 

We also have this code within `@hive/api`.

I would prefer to avoid `@hive/api` being a dependency of this migration.
E.g. `@hive/api` could change and then we would have to adjust the DB migrations accordingly.
I would rather prefer the migration to be a frozen"" standalone piece of that never needs to change.

But maybe I am being overcautious here? 😁